src/Features/Core/Portable/ConvertIfToSwitch/AbstractConvertIfToSwitchCodeRefactoringProvider.cs

```diff
@@ -1,256 +1,124 @@
 
 namespace Microsoft.CodeAnalysis.ConvertIfToSwitch
 {
-    internal abstract class AbstractConvertIfToSwitchCodeRefactoringProvider : CodeRefactoringProvider
+    internal abstract partial class AbstractConvertIfToSwitchCodeRefactoringProvider<
+        TIfStatementSyntax, TExpressionSyntax, TIsExpressionSyntax, TPatternSyntax> : CodeRefactoringProvider
     {
         public sealed override async Task ComputeRefactoringsAsync(CodeRefactoringContext context)
         {
             var document = context.Document;
+            var cancellationToken = context.CancellationToken;
+
             if (document.Project.Solution.Workspace.Kind == WorkspaceKind.MiscellaneousFiles)
             {
                 return;
             }
 
-            var syntaxFacts = document.GetLanguageService<ISyntaxFactsService>();
-            var semanticModel = await document.GetSemanticModelAsync().ConfigureAwait(false);
-
-            await CreateAnalyzer(syntaxFacts, semanticModel)
-                .ComputeRefactoringsAsync(context).ConfigureAwait(false);
-        }
-
-        protected interface IPattern<TSwitchLabelSyntax>
-            where TSwitchLabelSyntax : SyntaxNode
-        {
-            TSwitchLabelSyntax CreateSwitchLabel();
-        }
-
-        protected interface IAnalyzer
-        {
-            Task ComputeRefactoringsAsync(CodeRefactoringContext context);
-        }
-
-        protected abstract IAnalyzer CreateAnalyzer(ISyntaxFactsService syntaxFacts, SemanticModel semanticModel);
-
-        protected abstract class Analyzer<TStatementSyntax, TIfStatementSyntax, TExpressionSyntax, TSwitchLabelSyntax> : IAnalyzer
-            where TExpressionSyntax : SyntaxNode
-            where TIfStatementSyntax : SyntaxNode
-            where TSwitchLabelSyntax : SyntaxNode
-        {
-            protected readonly ISyntaxFactsService _syntaxFacts;
-            protected readonly SemanticModel _semanticModel;
-            private int _numberOfSubsequentIfStatementsToRemove = -1;
-            private TExpressionSyntax _switchExpression;
-            private Optional<TStatementSyntax> _switchDefaultBodyOpt;
-
-            public Analyzer(ISyntaxFactsService syntaxFacts, SemanticModel semanticModel)
+            var ifStatement = await context.TryGetRelevantNodeAsync<TIfStatementSyntax>().ConfigureAwait(false);
+            if (ifStatement == null || ifStatement.ContainsDiagnostics)
             {
-                _syntaxFacts = syntaxFacts;
-                _semanticModel = semanticModel;
+                return;
             }
 
-            public async Task ComputeRefactoringsAsync(CodeRefactoringContext context)
+            var semanticModel = await document.GetSemanticModelAsync(cancellationToken).ConfigureAwait(false);
+            var ifOperation = semanticModel.GetOperation(ifStatement);
+            if (!(ifOperation is IConditionalOperation { Parent: IBlockOperation { Operations: var operations } }))
```

Comment:
minor critique, and i'm not really opposed to this.  You do a very good job using patterns to drill into the smallest detail of what you need.  But i think it obfuscates slightly.  For example, if this was: `if (!(ifOperation is IConditionalOperation { Parent: IBlockOperation parentBlock }))` then subsequently we would see `parentBlock.operations` making it clear what `operations` were being talked about.  I feel like this sort of deconstruciton you have forces a bit more to remember how everything was computed when seeing it later on.

Reply:
note: i don't know what the right balance is :)

Reply:
I just didn't inline it because I *have to* inline span (can't define a span local here) and it gets kinda long. perhaps a local for operations would be better.